Source/inv.cpp

```diff
@@ -652,110 +652,67 @@ int SpecialAutoPlace(int pnum, int ii, int sx, int sy, int saveflag)
-					++plr[v1]._pNumInv;
-					*v10 = plr[v1]._pNumInv;
-					v12 = plr[v1].HoldItem._ivalue;
-					if (v12 < 2500) {
-						if (v12 > 1000)
-							*(int *)((char *)&plr[0].InvList[0]._iCurs + v11) = 5;
-						else
-							*(int *)((char *)&plr[0].InvList[0]._iCurs + v11) = 4;
-					} else {
-						*(int *)((char *)&plr[0].InvList[0]._iCurs + v11) = 6;
-					}
-					plr[v1]._pGold = CalculateGold(pnuma);
-					v3 = 1;
-				}
-				--v9;
-			} while (v9 >= 0);
-		} else {
-			v7 = &plr[v1].InvList[0]._ivalue;
-			while (!v3) {
-				if (*(v7 - 47) == 11 && *v7 < 5000) {
-					v8 = plr[v1].HoldItem._ivalue + *v7;
-					if (v8 <= 5000) {
-						*v7 = v8;
-						if (v8 < 2500) {
-							if (v8 > 1000)
-								*(v7 - 1) = 5;
-							else
-								*(v7 - 1) = 4;
-						} else {
-							*(v7 - 1) = 6;
-						}
-						plr[v1]._pGold = CalculateGold(pnuma);
-						v3 = 1;
-					}
-				}
-				++v6;
-				v7 += 92;
-				if (v6 >= plr[v1]._pNumInv) {
-					if (v3)
-						return v3;
-					goto LABEL_28;
-				}
+	BOOL done;
+	int i, ii;
+	int xx, yy;
+
+	done = FALSE;
+	for (i = 0; i < plr[pnum]._pNumInv && !done; i++) {
+		if (plr[pnum].InvList[i]._itype == ITYPE_GOLD) {
+			if (plr[pnum].HoldItem._ivalue + plr[pnum].InvList[i]._ivalue <= 5000) {
+				plr[pnum].InvList[i]._ivalue = plr[pnum].HoldItem._ivalue + plr[pnum].InvList[i]._ivalue;
+				if (plr[pnum].InvList[i]._ivalue >= 2500)
+					plr[pnum].InvList[i]._iCurs = ICURS_GOLD_LARGE;
+				else if (plr[pnum].InvList[i]._ivalue <= 1000)
+					plr[pnum].InvList[i]._iCurs = ICURS_GOLD_SMALL;
+				else
+					plr[pnum].InvList[i]._iCurs = ICURS_GOLD_MEDIUM;
+				plr[pnum]._pGold = CalculateGold(pnum);
+				done = TRUE;
 			}
 		}
-	} else {
-		v4 = &plr[v1].InvList[0]._ivalue;
-		while (!v3) {
-			if (*(v4 - 47) == 11) {
-				v5 = *v4 + plr[v1].HoldItem._ivalue;
-				if (v5 <= 5000) {
-					*v4 = v5;
-					if (v5 < 2500) {
-						if (v5 > 1000)
-							*(v4 - 1) = 5;
-						else
-							*(v4 - 1) = 4;
-					} else {
-						*(v4 - 1) = 6;
-					}
-					plr[v1]._pGold = CalculateGold(pnuma);
-					v3 = 1;
-				}
-			}
-			++v2;
-			v4 += 92;
-			if (v2 >= plr[v1]._pNumInv) {
-				if (v3)
-					return v3;
-				goto LABEL_14;
+	}
+	if (done)
+		return done;
+
+	for (i = 0; i < plr[pnum]._pNumInv && !done; i++) {
+		if (plr[pnum].InvList[i]._itype == ITYPE_GOLD && plr[pnum].InvList[i]._ivalue < 5000) {
+			if (plr[pnum].HoldItem._ivalue + plr[pnum].InvList[i]._ivalue <= 5000) {
+				plr[pnum].InvList[i]._ivalue = plr[pnum].HoldItem._ivalue + plr[pnum].InvList[i]._ivalue;
+				if (plr[pnum].InvList[i]._ivalue >= 2500)
+					plr[pnum].InvList[i]._iCurs = ICURS_GOLD_LARGE;
+				else if (plr[pnum].InvList[i]._ivalue <= 1000)
+					plr[pnum].InvList[i]._iCurs = ICURS_GOLD_SMALL;
+				else
+					plr[pnum].InvList[i]._iCurs = ICURS_GOLD_MEDIUM;
+				plr[pnum]._pGold = CalculateGold(pnum);
+				done = TRUE;
 			}
 		}
 	}
-	return v3;
+	if (done)
+		return done;

```

Comment:
You can remove this return since the following for loop checks `!done` and will therefore skip the loop and do the same thing.

Reply:
It won't be bin exact as it would then check both the following loops and not do an early return instead.

Reply:
Correct, since the loop has `!done` the compiler will see that and skip it. It should still be bin exact.

Reply:
It isn't :)

Reply:
The reason it won't be bin exact without this `if done return` is that the loop has an initial assignment statement `i = 39` that is always executed, regardless of whether `done` is true or not.

Reply:
Makes sense. I suppose if the init value was `0` instead, it'd work in that case.